text/0097-log-data-model.md

```diff
@@ -0,0 +1,971 @@
+
+- Ability to efficiently represent named top-level fields, which are almost
+  always present (e.g. when using encodings like Protocol Buffers where fields
+  are enumerated but not named on the wire).
+  
+- Ability to enforce types of named fields, which is very useful for compiled
+  languages with type checks.
+
+- Flexibility to represent less frequent data via key/value pair lists. This
+  includes well-known data that has standardized semantics as well as arbitrary
+  custom data that the application may want to include in the logs.
+
+When designing this data model I followed the following reasoning to make a
+decision about when to use use a top-level named field:
+
+- The field needs to be either mandatory for all records or be frequently
+  present in well-known log and event formats (such as `Timestamp`) or is
+  expected to be often present in log records in upcoming logging systems (such
+  as `TraceId`).
+
+- The field’s semantics must be the same for all known log and event formats and
+  can be mapped directly and unambiguously to this data model.
+  
+Both of the above conditions were required to give the field a place in the
+top-level structure of the record.
+
+## Log and Event Record Definition
+
+Note: below we use type `any`, which can be a scalar value (number, string or
+boolean), or an array or map of values. Arbitrary deep nesting of values for
+arrays and maps is allowed (essentially allow to represent an equivalent of a
+JSON object).
+
+[Appendix A](#appendix-a-example-mappings) contains many examples that show how
+existing log formats map to the fields defined below. If there are questions
+about the meaning of the field reviewing the examples may be helpful.
+
+Here is the list of fields in a log record:
+
+Field Name     |
+---------------|
+Timestamp      |
+TraceId        |
+SpanId         |
+TraceFlags     |
+SeverityText   |
+SeverityNumber |
+ShortName      |
+Body           |
+Resource       |
+Attributes     |
+
+Below is the detailed description of each field.
+
+### Field: `Timestamp`
+
+Type: Timestamp, uint64 nanoseconds since Unix epoch.
+
+Description: Time when the event occurred measured by the origin clock. This
+field is optional, it may be missing the timestamp is unknown.
+
+### Trace Context Fields
+
+#### Field: `TraceId`
+
+Type: byte sequence.
+
+Description: Optional request trace id. Can be set for logs that are part of
+request processing and have an assigned trace id.
+
+#### Field: `SpanId`
+
+Type: byte sequence.
+
+Description: Optional span id. Can be set for logs that are part of a particular
+processing span. If SpanId is present TraceId SHOULD be also present.
+
+#### Field: `TraceFlags`
+
+Type: byte.
+
+Description: Optional trace flag as defined in
+[W3C trace context](https://www.w3.org/TR/trace-context/#trace-flags)
+specification. At the time of writing the specification defines one flag - the
+SAMPLED flag.
+
+### Severity Fields
```

Comment:
I am not sure of the pros & cons of having `SeverityText`  and `SeverityNumber` as a top-level fields when log messages are commonly more characterized than just the severity/level, also characterized with a `logger.name` and sometimes a `logger.thread`. 
`logger.thread` being particularly useful to correlate log messages together when not using distributed traces (adoption path of OpenTelemetry).

I see benefits in
* defining an extensible logging namespace (`log.severityText`, `log.severityNumber`, `log.name`, `log.thread`, `log.origin.file.name`, `log.origin.file.line`, `class`, etc)
* grouping all the logging fields together. 
   * Grouping logging fields together suggests that we would either move logging fields under `attributes` or introduce a top-level grouping field `log`

Sample group of log fields

```
├── log
│   ├── level: INFO
│   ├── logger: com.mycie.donut.shop
│   └── thread: http-123
```

Reply:
>I am not sure of the pros & cons of having SeverityText and SeverityNumber as a top-level fields when log messages are commonly more characterized than just the severity/level, also characterized with a logger.name and sometimes a logger.thread.

The Design Notes section clarifies the cons. Having `Severity*` as top-level fields does not prevent from recording additional (arguably more rarely used) fields in the `Attributes`.

>logger.thread being particularly useful to correlate log messages together when not using distributed traces (adoption path of OpenTelemetry).

I am not able to find `logger.thread` under [log.*](https://www.elastic.co/guide/en/ecs/current/ecs-log.html) in ECS. Is there a currently published definition? If not please expand on what it means.

>defining an extensible logging namespace

What does "extensible" mean? Is this different than extensibility that is possible by having key/values in `Attributes` of this data model?

>I see benefits in
>...
>grouping all the logging fields together. 

What are the benefits? What is the selection criteria for the "logging fields" grouping that you suggest? I am afraid it is not obvious to me.

Reply:
> The Design Notes section clarifies the cons. Having `Severity*` as top-level fields does not prevent from recording additional (arguably more rarely used) fields in the Attributes.

understood.

> I am not able to find `logger.thread` under log.* in ECS. Is there a currently published definition? If not please expand on what it means.


you are right, I made a mistake, we use the `process.thread.name` from the [process](https://www.elastic.co/guide/en/ecs/current/ecs-process.html) namespace to indicate the thread that emitted the message.

It would look like:

```javascript
{"Attributes":{
      "log.level":"INFO",
      "log.logger":"com.mycie.donut.shop",
      "process.thread.name":"http-123"
}}
```


> What does "extensible" mean? Is this different than extensibility that is possible by having key/values in Attributes of this data model?

My phrasing may have been awkward. I meant that there are a bunch of information we may want to add to characterise a log message.

Here are some examples 

Sample log message coming from a java application

```javascript
{"Attributes":{
      "log.level":"INFO",
      "log.logger":"com.mycie.donut.shop",
      "log.origin.file.name":"OrderManager.java",
      "log.origin.file.line":" 123",
      "log.origin.function":"placeOrder",

      "process.thread.name":"http-123"
}}
```

Sample exception log message coming from a java application


```javascript
{"Attributes": {
    "log.level": "INFO", 
    "log.logger": "com.mycie.donut.shop", 
    "log.origin.file.name": "OrderManager.java", 
    "log.origin.file.line": "123", 
    "log.origin.function": "placeOrder", 

    "error.type": "InvalidToppingException", 
    "error.message": "Cannot put tomato sauce on a donut", 
    "error.stack_trace": "TODO add multi line stack trace", 
 
    "process.thread.name": "http-123"
}}
```


Sample log message coming from a java application and collected from a log file

```javascript
{"Attributes":{
      "log.level":"INFO",
      "log.logger":"com.mycie.donut.shop",
      "log.origin.file.name":"OrderManager.java",
      "log.origin.file.line":" 123",
      "log.origin.function":"placeOrder",

      "process.thread.name":"http-123",

      "log.file.path": "/var/log/donut_shop/application.log"
   }
   // ...
}
```

Sample log message coming from a web server access log

{
   "Attributes":{
      "http.scheme":"https",
      "http.host":"donut.mycie.com",
      "http.target":"/order",
      "http.method":"post",
      "http.status_code":500,
      "http.client_ip":"192.0.2.5",
      "http.referer":"-",
      "http.flavor":"1.1",
      "http.response.bytes":542,
      "user_agent.name":"Chrome",
      "user_agent.version":"70.0.3538.102",
      "user_agent.os.name":"Mac OS X",
      "user_agent.os.version":"10.14.0",
      "user_agent.original":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36"

      "log.file.path": "/var/log/nginx/access.log"
   }
}

> What are the benefits? What is the selection criteria for the "logging fields" grouping that you suggest? I am afraid it is not obvious to me.

I was referring to my idea that visualization benefits from a hierarchical visualisation of the data structure. You explained in another comment that for consistency with metrics and traces, we preferred a "flat list with dots". I support your point that consistency with traces and metrics is more important than this visualization comfort question.

Reply:
I second the comments by @cyrille-leclerc ....  if you look at the logging levels and naming nomenclature in different languages (Java/JVM, Python, etc.) it is clear there is a really wide range, and boiling it down to two fields likely would not suffice.  

We could have a required logging severity field, which is text, but a numeric one likely would be really difficult to standardize on for every language.

Having a defined group (or map / key-value attribute list) specifically for logging severity metadata would be really useful.  Putting the same data in attributes would pollute the attributes.....  and make it harder to standardize I would think.

Reply:
>you are right, I made a mistake, we use the process.thread.name from the process namespace to indicate the thread that emitted the message.

@cyrille-leclerc Thank you. I now understand the meaning of the "process.thread.name" but I still don’t see the need to make it a top-level field or tie to severity level in anyway. "process.thread.name" is a field that can live nicely in the `Attributes`. 

>My phrasing may have been awkward. I meant that there are a bunch of information we may want to add to characterise a log message.

That’s fine. That is what `Attributes` are for. Anything _in addition_ to what this data model defines can be in the `Attributes`. That is the intent and nothing prevents adding "process.thread.name", "log.origin.function" or "error.stack_trace" to `Attributes`.

Reply:
@cyrille-leclerc I added the 3rd example that is based on the web server example you posted but uses OpenTelemetry "http.*" semantic conventions.

Reply:
Resolving this as it appears I addressed all questions either here or in other replies in this PR. Please feel free to reopen if I missed something.